ortools/sat/timetable_horizontal_edgefinding.cc

```diff
@@ -0,0 +1,290 @@
+
+#include "absl/log/check.h"
+#include "ortools/base/iterator_adaptors.h"
+#include "ortools/sat/integer.h"
+#include "ortools/sat/intervals.h"
+#include "ortools/sat/model.h"
+#include "ortools/util/sort.h"
+#include "ortools/util/strong_integers.h"
+
+namespace operations_research {
+namespace sat {
+
+HorizontallyElasticOverloadChecker::HorizontallyElasticOverloadChecker(
+    AffineExpression capacity, SchedulingConstraintHelper* helper,
+    SchedulingDemandHelper* demands, Model* model)
+    : num_tasks_(helper->NumTasks()),
+      capacity_(capacity),
+      helper_(helper),
+      demands_(demands),
+      integer_trail_(model->GetOrCreate<IntegerTrail>()) {
+  task_types_.resize(num_tasks_);
+  profile_events_.reserve(4 * num_tasks_ + 1);
+}
+
+void HorizontallyElasticOverloadChecker::RegisterWith(
+    GenericLiteralWatcher* watcher) {
+  const int id = watcher->Register(this);
+  watcher->WatchUpperBound(capacity_, id);
+  helper_->WatchAllTasks(id, watcher);
+  for (int t = 0; t < num_tasks_; t++) {
+    watcher->WatchLowerBound(demands_->Demands()[t], id);
+  }
+  watcher->SetPropagatorPriority(id, 3);
+  watcher->NotifyThatPropagatorMayNotReachFixedPointInOnePass(id);
+}
+
+bool HorizontallyElasticOverloadChecker::Propagate() {
+  if (!helper_->SynchronizeAndSetTimeDirection(true)) return false;
+  if (!OverloadCheckerPass()) return false;
+  if (!helper_->SynchronizeAndSetTimeDirection(false)) return false;
+  if (!OverloadCheckerPass()) return false;
+  return true;
+}
+
+bool HorizontallyElasticOverloadChecker::OverloadCheckerPass() {
+  // Prepate the profile events which will be used during ScheduleTasks to
+  // dynamically compute the profile. The events are valid for the entire
+  // function and do not need to be recomputed. 
+  //
+  // TODO: This datastructure contains everything we need to compute the 
+  // "classic" profile used in Time-Tabling. 
+  profile_events_.clear();
+  for (int i = 0; i < num_tasks_; i++) {
+    if (helper_->IsPresent(i) && demands_->DemandMin(i) > 0) {
+      profile_events_.emplace_back(i, helper_->StartMin(i),
+                                   demands_->DemandMin(i),
+                                   ProfileEventType::kStartMin);
+      profile_events_.emplace_back(i, helper_->StartMax(i),
+                                   demands_->DemandMin(i),
+                                   ProfileEventType::kStartMax);
+      profile_events_.emplace_back(i, helper_->EndMin(i),
+                                   demands_->DemandMin(i),
+                                   ProfileEventType::kEndMin);
+      profile_events_.emplace_back(i, helper_->EndMax(i),
+                                   demands_->DemandMin(i),
+                                   ProfileEventType::kEndMax);
+    }
+  }
+  profile_events_.emplace_back(-1, kMaxIntegerValue, IntegerValue(0),
+                               ProfileEventType::kSentinel);
+  IncrementalSort(profile_events_.begin(), profile_events_.end());
```

Comment:
I think that if you are not updating `profile_events_` incrementally there is no point on sorting it with this method. I think plain `std::sort` would be cleaner.

Reply:
You're right, changed the code to truly build the profile incrementally by updating the events rather than building it from scratch each time. 

I don't think this drives much benefits yet but it opens the door for further incremental optimization using reversibles similarly to what is done in time-table.